### PR TITLE
Allow NME to compile WebGPU shaders

### LINKED ISSUE #14929: Add wgsl support to nme
- [x] Add a button to switch between webgl and webgpu on nme
- [x] useLogarithmicDepth
- [x] Live update from NME

List of nodes:

- [x] AnisotropyBlock
- [x] ClearCoatBlock
- [x] ReflectionBlock
- [x] RefractionBlock
- [x] SheenBlock
- [x] SubSurfaceBlock
- [x] PBRMetallicRoughnessBlock
- [x] ShadowMapBlock
- [x] PrePassTextureBlock
- [x] PrePassOutputBlock
- [x] ReflectionTextureBlock
- [x] ParticlePositionWorldBlock
- [x] ParticleTextureMaskBlock
- [x] ParticleTextureBlock
- [x] ParticleUVBlock
- [x] ParticleBlendMultiplyBlock
- [x] ParticleRampGradientBlock
- [x] ParticleColorBlock
- [x] ImageProcessingBlock
- [x] SceneDepthBlock
- [x] ScreenPositionBlock
- [x] CurrentScreenBlock
- [x] ScreenSpaceBlock
- [x] ScreenSizeBlock
- [x] HeightToNormalBlock
- [x] TBNBlock
- [x] PerturbNormalBlock
- [x] MorphTargetsBlock
- [x] BonesBlock
- [x] ViewDirectionBlock
- [x] LightInformationBlock
- [x] InstanceColorBlock
- [x] ClipPlanesBlock
- [x] InstancesBlock
- [x] FragmentOutputBlock
- [x] FragCoordBlock
- [x] MatrixBuilderBlock
- [x] BiPlanarBlock
- [x] ImageSourceBlock
- [x] TriPlanarBlock
- [x] FragDepthBlock
- [x] DiscardBlock
- [x] MaterialAlphaBlock
- [x] CameraPositionBlock
- [x] CameraParametersBlock
- [x] FogColorBlock
- [x] FogBlock
- [x] RemapBlock
- [x] ClampBlock
- [x] NormalizeBlock
- [x] LightBlock
- [x] MatrixTransposeBlock
- [x] NormalBlendBlock
- [x] MatrixIndicesBlock
- [x] MatrixWeightsBlock
- [x] MatrixIndicesExtraBlock
- [x] MatrixWeightsExtraBlock
- [x] FrontFacingBlock
- [x] MeshAttributeExistsBlock
- [x] RandomNumberBlock
- [x] SimplexPerlin3DBlock
- [x] WorleyNoise3DBlock
- [x] CloudBlock
- [x] VoronoiNoiseBlock
- [x] WorldTangentBlock
- [x] WorldNormalBlock
- [x] WorldPositionBlock
- [x] ReplaceColorBlock
- [x] PosterizeBlock
- [x] GradientBlock
- [x] DesaturateBlock
- [x] ColorMergerBlock
- [x] ColorSplitterBlock
- [x] VectorMergerBlock
- [x] VectorSplitterBlock
- [x] Float
- [x] Vector2
- [x] Vector3
- [x] Vector4
- [x] Color3
- [x] Color4
- [x] TextureBlock
- [x] TimeBlock
- [x] RealTimeBlock
- [x] DeltaTimeBlock
- [x] LerpBlock
- [x] StepBlock
- [x] SmoothStepBlock
- [x] NLerpBlock
- [x] EqualBlock
- [x] NotEqualBlock
- [x] LessThanBlock
- [x] LessOrEqualBlock
- [x] GreaterThanBlock
- [x] GreaterOrEqualBlock
- [x] XorBlock
- [x] OrBlock
- [x] AndBlock
- [x] AddBlock
- [x] DivideBlock
- [x] MaxBlock
- [x] MinBlock
- [x] ModBlock
- [x] MultiplyBlock
- [x] NegateBlock
- [x] OneMinusBlock
- [x] ReciprocalBlock
- [x] ScaleBlock
- [x] SignBlock
- [x] SqrtBlock
- [x] SubtractBlock
- [x] AbsBlock
- [x] ArcCosBlock
- [x] ArcSinBlock
- [x] ArcTanBlock
- [x] ArcTan2Block
- [x] CosBlock
- [x] CurveBlock
- [x] DegreesToRadiansBlock
- [x] ExpBlock
- [x] Exp2Block
- [x] FractBlock
- [x] LogBlock
- [x] PowBlock
- [x] RadiansToDegreesBlock
- [x] SawToothWaveBlock
- [x] SinBlock
- [x] SquareWaveBlock
- [x] TanBlock
- [x] TriangleWaveBlock
- [x] CrossBlock
- [x] DerivativeBlock
- [x] DistanceBlock
- [x] DotBlock
- [x] FresnelBlock
- [x] LengthBlock
- [x] ReflectBlock
- [x] RefractBlock
- [x] Rotate2dBlock
- [x] TransformBlock
- [x] TwirlBlock
- [x] Matrix
- [x] WorldMatrixBlock
- [x] WorldViewMatrixBlock
- [x] WorldViewProjectionMatrixBlock
- [x] ViewMatrixBlock
- [x] ViewProjectionMatrixBlock
- [x] ProjectionMatrixBlock
- [x] MatrixDeterminantBlock
- [x] ElbowBlock
- [x] TeleportInBlock
- [x] TeleportOutBlock
- [x] PositionBlock
- [x] UVBlock
- [x] ColorBlock
- [x] NormalBlock
- [x] TangentBlock
- [x] VertexOutputBlock
- [x] RoundBlock
- [x] CeilingBlock
- [x] FloorBlock

### PR DESCRIPTION
**Pull request by deltakosh**
Fix #14929

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts [hunk 1]
```diff
@@ -192,7 +211,11 @@ export class NodeMaterialBuildState {
      */
     public _emit2DSampler(name: string) {
         if (this.samplers.indexOf(name) < 0) {
-            this._samplerDeclaration += `uniform sampler2D ${name};\n`;
+            if (this.shaderLanguage === ShaderLanguage.WGSL) {
+                this._samplerDeclaration += `var ${name}: sampler;\n`;
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts:215]: You will also have to generate a texture object, so something like:
```glsl
this._samplerDeclaration += `var ${name}Sampler: sampler;\n`;
this._samplerDeclaration += `var ${name}Texture: texture_2d<f32>;\n`;
```
It assumes the type is `f32`, but I don't think we can use other types in the NME (yet?).

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14916/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14916/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14916/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14916/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14916/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14916/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14916/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14916/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14916/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14916/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts [hunk 2]
```diff
@@ -538,4 +538,74 @@ export class NodeMaterialBuildState {
...
+    private _convertVariableDeclarationToWGSL(type: string, dest: string, source: string): string {
+        return source.replace(new RegExp(`(${type})\\s+(\\w+)`, "g"), `var $2: ${dest}`);
+    }
+
+    private _convertVariableConstructorsToWGSL(type: string, dest: string, source: string): string {
+        return source.replace(new RegExp(`(${type})\\(`, "g"), ` ${dest}(`);
+    }
+
+    private _convertOutParametersToWGSL(source: string): string {
+        return source.replace(new RegExp(`out\\s+var\\s+(\\w+):\\s+(\\w+)`, "g"), `$1: ptr<function, $2>`);
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts:551]: Maybe you should allow optional whitespaces after the identifier name and before the semi-colon, and also allow no whitespaces after the semi-colon?

Like:
```typescript
new RegExp(`out\\s+var\\s+(\\w+)\\s*:\\s*(\\w+)`, "g")`
```

File: packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts [hunk 3]
```diff
@@ -538,4 +538,74 @@ export class NodeMaterialBuildState {
...
+    }
+
+    private _convertFunctionsToWGSL(source: string): string {
+        const regex = /var\s+(\w+):\s+(\w+)\((.*)\)/;
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts:555]: Same than above.

File: packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts [hunk 4]
```diff
@@ -538,4 +538,74 @@ export class NodeMaterialBuildState {
...
+            // Constructing the final output string
+            return source.replace(match[0], `fn ${funcName}(${formattedParams}) -> ${funcType}`);
+        }
+        return source;
+    }
+
+    public _babylonSLtoWGSL(code: string) {
+        // variable declarations
+        code = this._convertVariableDeclarationToWGSL("void", "voidnull", code);
+        code = this._convertVariableDeclarationToWGSL("int", "i32", code);
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts:575]: Also:
```typescript
code = this._convertVariableDeclarationToWGSL("uint", "u32", code);
```

File: packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts (hunk 2 above)
Comment (deltakosh) [reply packages/dev/core/src/Materials/Node/nodeMaterialBuildState.ts:551]: I agree that will make it more flexible

**Review by Popov72 — APPROVED** [review]

**Pull request merged**